src/widgets/panel/panel.tsx

```diff
@@ -76,18 +76,30 @@ export type PanelTemplates = {
   collapseButtonText?: Template<{ collapsed: boolean }>;
 };
 
-export type PanelWidgetOptions = {
+export type PanelWidgetOptions<TWidget extends WidgetFactory<any, any, any>> = {
   /**
    * A function that is called on each render to determine if the
    * panel should be hidden based on the render options.
    */
-  hidden?(options: RenderOptions): boolean;
+  hidden?(
+    options: RenderOptions & {
+      widgetRenderState: ReturnType<
+        Exclude<ReturnType<TWidget>['getWidgetRenderState'], undefined>
```

Comment:
Hmm I'm not sure what line 87 does then, it looks really complicated.

Reply:
So, Let me break it down.

First of all, the usage would be

```js
panel<typeof breadcrumb>({ ... })(breadcrumb);
```

And

> `TWidget extends WidgetFactory`

- It means TWidget is a WidgetFactory (Yeah it's TWidget, but not widget, but a widget factory).
- `ReturnType<Twidget>` -> It returns a widget
- `ReturnType<TWidget>['getWidgetRenderState']` -> From the widget, I'm picking `getWidgetRenderState` function
- `getWidgetRenderState` can exist, but in case of custom widget, it might be `undefined`.
- `Exclude<ReturnType<TWidget>['getWidgetRenderState'], undefined>` -> However we assume it exists here, so picking the real type.
- `ReturnType<Exclude<ReturnType<TWidget>['getWidgetRenderState'], undefined>>` -> Now we've got the return of `getWidgetRenderState`, which is going to be one of SearchBox WidgetRenderState, Autocomplete WidgetRenderState, Breadcrumb WidgetRenderState, ...

Yeah that long journey is to get the real WidgetRenderState for it, but not sure how to clean it up.

Reply:
```ts
export type PanelRenderOptions<
  TWidget extends WidgetFactory<any, any, any>
> = RenderOptions & {
  widgetRenderState: ReturnType<TWidget>['getWidgetRenderState'] extends (
    renderOptions: any
  ) => infer TRenderState
    ? TRenderState
    : unknown;
};
```

I think this is a little simpler maybe?